pootle/apps/pootle_store/diff.py

```diff
@@ -0,0 +1,222 @@
+        # but do not appear in the file
+        new_units = [u for u in self.updated_db_units
+                     if u not in self.file_unit_ids]
+
+        # These unit are either present in both or only in the file so are
+        # kept in the file order
+        new_units += [u for u in self.file_unit_ids
+                      if u not in self.obsoleted_db_units]
+
+        return new_units
+
+    @cached_property
+    def obsoleted_db_units(self):
+        return [unitid for unitid, unit in self.db_units.items()
+                if unit['state'] == OBSOLETE
+                and unit["revision"] > self.file_revision]
+
+    @cached_property
+    def opcodes(self):
+        sm = difflib.SequenceMatcher(None, self.active_units,
+                                     self.new_unit_list)
+        return sm.get_opcodes()
+
+    @cached_property
+    def updated_db_units(self):
+        return [unitid for unitid, unit in self.db_units.items()
+                if unit['revision'] > self.file_revision
+                and not unit["state"] == OBSOLETE]
+
+    def get_updated_dbids(self):
+        """Returns a set of unit DB ids to be updated.
+        :return: a set of unit DB ids to be updated.
+        """
+        update_dbids = set()
+
+        def _has_changed(uid):
+            """ check if the source and targets match
+            or are equal to empty string
+            """
+            file_unit = self.file_units[uid]
+            db_unit = self.db_units[uid]
+
+            db_source = multistring_to_python(db_unit["source_f"])
+            db_target = multistring_to_python(db_unit["target_f"])
+
+            source_same = ((file_unit["source"] == "" and db_source == "") or
+                           file_unit["source"] == db_source)
+            target_same = ((file_unit["target"] == "" and db_target == "") or
+                           file_unit["target"] == db_target)
+
+            if source_same and target_same:
+                return False
+            return True
+
+        for (tag, i1, i2, j1, j2) in self.opcodes:
+            if tag == 'equal':
+                dbids = set(self.db_units[uid]['id']
+                            for uid in self.active_units[i1:i2]
+                            if uid in self.file_unit_ids
+                            and _has_changed(uid))
+                update_dbids.update(dbids)
+        return update_dbids
+
+    def diff(self):
+        """Return a dictionary of change actions"""
+        _diff = {}
+        _diff["index"] = self.get_indexes_to_update()
+        _diff["obsolete"] = self.get_units_to_obsolete()
+        _diff["add"] = self.get_units_to_add()
+        _diff["update"] = self.get_units_to_update()
+        return _diff
+
+    def get_units_to_obsolete(self):
+        return [unit['id'] for unitid, unit in self.db_units.items()
+                if unitid not in self.file_unit_ids
+                and unitid in self.active_units
```

Comment:
Adding my 2c.  If there is a contradiction in the styleguide then lets clean it up please.  Most style is a choice and the readability comes from consistent application.  If it's not working then lets discuss one aspect and fix it.

Note the Pootle code-base is old.  So don't measure inconstancy as not complying with the style guide.  Just that that code hasn't got cleaned up yet.

In terms of PEP8.  It really isn't the be all and end all.  I like my code to comply.  But even in PEP8 there are mutually exclusive options as well as options that make code less readable, so we have to make choices.  As far as PEP8 goes now we do comply fully to the choices made in ``pootle/tools/pep8.sh``.  We fail on Travis if any of these PEP8 requirements fail ``E10,E11,E222,E251,E261,E262,E27,E401,E70,E711,E712,E713,E721,W191,W291,W292,W293,W391,W60``.

Also not that we haven't applied pep8 to tests, they where far too messy, but we could expand our coverage there if needed.

We can always expand the list of checks that we require to pass, but that would mean:

1. Cleaning up cases in the code to get 100% pass for that PEP8 test
2. Making a choice where there is a conflict of options
3. Adding it to ``pep8.sh`` so that we catch these in Travis

@phlax if you have ones that continue to flag in your editor (as I assume its base choice is not conflicting) feel free to report an issue as I'm a happy PEP8 cleaner.